File: frontend/src/api/bookmarks.ts
import { api } from './client'

export const bookmarksApi = {
  /** 获取书签列表 (as_tree=true 返回树形结构) */
  list: (asTree = true) => api.get('/api/bookmarks/', { params: { as_tree: asTree } }),
  getBookmarks: (asTree = true) => api.get('/api/bookmarks/', { params: { as_tree: asTree } }),

  /** 创建书签/文件夹 */
  create: (data: any) => api.post('/api/bookmarks/', data),
  createBookmark: (data: any) => api.post('/api/bookmarks/', data),

  /** 更新书签 */
  update: (id: string, data: any) => api.put(`/api/bookmarks/${id}`, data),
  updateBookmark: (id: string, data: any) => api.put(`/api/bookmarks/${id}`, data),

  /** 删除书签 */
  delete: (id: string) => api.delete(`/api/bookmarks/${id}`),
  deleteBookmark: (id: string) => api.delete(`/api/bookmarks/${id}`),

  /** 清空所有书签 */
  deleteAll: () => api.delete('/api/bookmarks/'),
  clearBookmarks: () => api.delete('/api/bookmarks/'),

  /** 重复检测 */
  getDuplicates: () => api.get('/api/bookmarks/duplicates'),
  findDuplicates: () => api.get('/api/bookmarks/duplicates'),

  /** 批量健康检查 */
  checkHealth: (urls: string[]) => api.post('/api/bookmarks/check-health', { urls }),

  /** 导出书签 HTML */
  export: () => api.get('/api/bookmarks/export', { responseType: 'blob' }),
  exportBookmarks: () => {
    const link = document.createElement('a')
    link.href = '/api/bookmarks/export'
    link.setAttribute('download', '')
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
  },

  /** 重新排序 */
  reorder: (orderedIds: string[], parentId: string | null = undefined) =>
    api.post('/api/bookmarks/reorder', { ordered_ids: orderedIds, parent_id: parentId }),
  reorderBookmarks: (orderedIds: string[], parentId: string | null = undefined) =>
    api.post('/api/bookmarks/reorder', { ordered_ids: orderedIds, parent_id: parentId }),

  /** AI 自动整理 (SSE 流式) */
  aiAutoOrganize: (data: any) => api.post('/api/bookmarks/ai-auto-organize', data),

  /** 导入 HTML 书签文件 */
  importHtml: (file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    return api.post('/api/bookmarks/import-html', formData)
  },

  /** 获取网站图标 */
  fetchIcon: (url: string) => api.get('/api/navigation/fetch-icon', { params: { url } }),
}

File: backend/app/api/bookmarks.py
from fastapi import APIRouter, HTTPException, Query, UploadFile, File, Response, Body
from typing import List, Optional, Dict
from datetime import datetime
from app.services import bookmark_service as service
from app.schemas.bookmark import BookmarkCreate, BookmarkUpdate, BookmarkResponse

router = APIRouter()

@router.get("/duplicates")
async def get_duplicates():
    return service.find_duplicates()

@router.post("/check-health")
async def check_health(payload: Dict = Body(...)):
    urls = payload.get("urls", [])
    if not urls:
        return {}
    return await service.check_batch_health(urls)

# 使用标准的 / 路径，确保匹配稳定
@router.get("/", response_model=List[BookmarkResponse])
async def list_bookmarks(as_tree: bool = Query(True)):
    return service.list_bookmarks(as_tree=as_tree)

@router.post("/", response_model=BookmarkResponse)
async def create_bookmark(bookmark: BookmarkCreate):
    # 自动本地化远程图标
    bookmark_data = bookmark.dict()
    icon = bookmark_data.get("icon")
    if icon and icon.startswith(("http://", "https://")):
        from .navigation import get_and_cache_favicon
        local_path = await get_and_cache_favicon(icon)
        if local_path:
            bookmark_data["icon"] = local_path
            
    return service.add_bookmark(bookmark_data)

@router.put("/{bm_id}")
async def update_bookmark(bm_id: str, bookmark: BookmarkUpdate):
    success = service.update_bookmark(bm_id, bookmark.dict(exclude_unset=True))
    if not success:
        raise HTTPException(status_code=404, detail="Bookmark not found")
    return {"message": "Updated"}

@router.delete("/{bm_id}")
async def delete_bookmark(bm_id: str):
    success = service.delete_bookmark(bm_id)
    if not success:
        raise HTTPException(status_code=404, detail="Bookmark not found")
    return {"message": "Deleted"}

@router.delete("/")
async def clear_bookmarks_root():
    service.clear_bookmarks()
    return {"message": "All bookmarks cleared"}

@router.get("/export")
async def export_bookmarks():
    html_content = service.export_bookmarks_to_html()
    filename = f"lens_bookmarks_{datetime.now().strftime('%Y%m%d')}.html"
    return Response(
        content=html_content,
        media_type="text/html",
        headers={
            "Content-Disposition": f"attachment; filename={filename}"
        }
    )

@router.post("/reorder")
async def reorder_bookmarks(payload: dict):
    ordered_ids = payload.get("ordered_ids", [])
    parent_id = payload.get("parent_id")
    service.reorder_bookmarks(ordered_ids, parent_id)
    return {"message": "Reordered"}

@router.post("/ai-auto-organize")
async def ai_auto_organize(payload: dict = Body({})):
    """全自动流式整理书签 (支持指定文件夹)"""
    from app.services.bookmark_ai_service import BookmarkAIService
    from fastapi.responses import StreamingResponse
    
    target_folder_id = payload.get("folder_id")
    
    async def generate():
        async for msg in BookmarkAIService.run_auto_organize(target_folder_id):
            yield f"data: {msg}\n\n"
            
    return StreamingResponse(generate(), media_type="text/event-stream")

@router.post("/import-html")
async def import_bookmarks_html(file: UploadFile = File(...)):
    if not file.filename.endswith(('.html', '.htm')):
        raise HTTPException(status_code=400, detail="请上传 HTML 格式的书签文件")
    
    content = await file.read()
    html_text = ""
    for enc in ['utf-8-sig', 'utf-8', 'gbk', 'iso-8859-1']:
        try:
            html_text = content.decode(enc)
            break
        except Exception:
            continue
            
    if not html_text:
        raise HTTPException(status_code=400, detail="无法解析文件编码")
            
    count = service.import_from_html(html_text)
    return {"message": f"成功导入 {count} 个项目", "count": count}